File: packages/client-python/src/trickle/type_hash.py
"""Produce a stable hash for a function's type signature."""

from __future__ import annotations

import hashlib
import json
from typing import Any, Dict, List


def hash_type(args_type: Dict[str, Any], return_type: Dict[str, Any]) -> str:
    """Return the first 16 hex characters of the SHA-256 hash of the
    canonicalized ``(args_type, return_type)`` pair."""
    canonical = {
        "args": _canonicalize(args_type),
        "return": _canonicalize(return_type),
    }
    payload = json.dumps(canonical, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]


def _canonicalize(node: Any) -> Any:
    """Recursively sort all dict keys and union members to produce a
    deterministic representation."""
    if isinstance(node, dict):
        result: Dict[str, Any] = {}
        for key in sorted(node.keys()):
            value = node[key]
            result[key] = _canonicalize(value)
        # Sort union members for stability
        if result.get("kind") == "union" and "members" in result:
            result["members"] = sorted(
                result["members"],
                key=lambda m: json.dumps(m, sort_keys=True, separators=(",", ":")),
            )
        return result
    if isinstance(node, list):
        return [_canonicalize(item) for item in node]
    return node

File: packages/client-js/src/type-hash.ts
import { createHash } from 'crypto';
import { TypeNode } from './types';

/**
 * Canonicalize a TypeNode for deterministic hashing.
 * - Object properties are sorted alphabetically by key at all levels.
 * - Union members are sorted by their canonical string representation.
 */
function canonicalize(node: TypeNode): unknown {
  switch (node.kind) {
    case 'primitive':
      return { kind: 'primitive', name: node.name };

    case 'array':
      return { kind: 'array', element: canonicalize(node.element) };

    case 'object': {
      const sortedKeys = Object.keys(node.properties).sort();
      const properties: Record<string, unknown> = {};
      for (const key of sortedKeys) {
        properties[key] = canonicalize(node.properties[key]);
      }
      return { kind: 'object', properties };
    }

    case 'union': {
      const members = node.members
        .map(m => canonicalize(m))
        .sort((a, b) => JSON.stringify(a).localeCompare(JSON.stringify(b)));
      return { kind: 'union', members };
    }

    case 'function': {
      return {
        kind: 'function',
        params: node.params.map(canonicalize),
        returnType: canonicalize(node.returnType),
      };
    }

    case 'promise':
      return { kind: 'promise', resolved: canonicalize(node.resolved) };

    case 'map':
      return { kind: 'map', key: canonicalize(node.key), value: canonicalize(node.value) };

    case 'set':
      return { kind: 'set', element: canonicalize(node.element) };

    case 'tuple':
      return { kind: 'tuple', elements: node.elements.map(canonicalize) };

    case 'unknown':
      return { kind: 'unknown' };

    default:
      return { kind: 'unknown' };
  }
}

/**
 * Hash the combined args + return type into a deterministic 16-hex-char string.
 */
export function hashType(argsType: TypeNode, returnType: TypeNode): string {
  const canonical = JSON.stringify({
    args: canonicalize(argsType),
    ret: canonicalize(returnType),
  });

  return createHash('sha256').update(canonical).digest('hex').substring(0, 16);
}
